sda-admin/main.go

```diff
@@ -0,0 +1,460 @@
+}
+
+func versionUsage() {
+	usageText := `Usage: sda-admin version
+  Show the version information for sda-admin.
+`
+	fmt.Println(usageText)
+}
+
+func printVersion() {
+	fmt.Printf("sda-admin version %s\n", version)
+}
+
+func checkToken(token string) {
+	if err := helpers.CheckTokenExpiration(token); err != nil {
+		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func parseFlagsAndEnv() {
+	// Set up flags
+	flag.StringVar(&api_uri, "uri", "", "Set the URI for the SDA server (optional if API_HOST is set)")
+	flag.StringVar(&token, "token", "", "Set the authentication token (optional if ACCESS_TOKEN is set)")
+
+	// Custom usage message
+	flag.Usage = func() {
+		fmt.Println(usage)
+	}
+
+	// Parse global flags first
+	flag.Parse()
+
+	// Check environment variables if flags are not provided
+	if api_uri == "" {
+		api_uri = os.Getenv("API_HOST")
+		if api_uri == "" {
+			fmt.Println("Error: either -uri must be provided or API_HOST environment variable must be set.")
+			os.Exit(1)
+		}
+	}
+
+	if token == "" {
+		token = os.Getenv("ACCESS_TOKEN")
+		if token == "" {
+			fmt.Println("Error: either -token must be provided or ACCESS_TOKEN environment variable must be set.")
+			os.Exit(1)
+		}
+	}
+}
+
+func handleHelpCommand() {
+	if flag.NArg() > 1 {
+		switch flag.Arg(1) {
+		case "list":
+			handleHelpList()
+		case "file":
+			handleHelpFile()
+		case "dataset":
+			handleHelpDataset()
+		case "version":
+			versionUsage()
+			os.Exit(0)
+		default:
+			fmt.Printf("Unknown command '%s'.\n", flag.Arg(1))
+			flag.Usage()
+		}
+	} else {
+		flag.Usage()
+	}
+	os.Exit(0)
+}
+
+func handleHelpList() {
+	if flag.NArg() == 2 {
+		listUsage()
+	} else if flag.NArg() > 2 && flag.Arg(2) == "users" {
+		listUsersUsage()
+	} else if flag.NArg() > 2 && flag.Arg(2) == "files" {
+		listFilesUsage()
+	} else {
+		fmt.Printf("Unknown subcommand '%s' for '%s'.\n", flag.Arg(2), flag.Arg(1))
+		listUsage()
+	}
+}
+
+func handleHelpFile() {
+	if flag.NArg() == 2 {
+		fileUsage()
+	} else if flag.NArg() > 2 && flag.Arg(2) == "ingest" {
+		fileIngestUsage()
+	} else if flag.NArg() > 2 && flag.Arg(2) == "accession" {
+		fileAccessionUsage()
+	} else {
+		fmt.Printf("Unknown subcommand '%s' for '%s'.\n", flag.Arg(2), flag.Arg(1))
+		fileUsage()
+	}
+}
+
+func handleHelpDataset() {
+	if flag.NArg() == 2 {
+		datasetUsage()
+	} else if flag.NArg() > 2 && flag.Arg(2) == "create" {
+		datasetCreateUsage()
+	} else if flag.NArg() > 2 && flag.Arg(2) == "release" {
+		datasetReleaseUsage()
+	} else {
+		fmt.Printf("Unknown subcommand '%s' for '%s'.\n", flag.Arg(2), flag.Arg(1))
+		datasetUsage()
+	}
+}
+
+func handleListCommand() {
+	if flag.NArg() < 2 {
+		fmt.Println("Error: 'list' requires a subcommand (users, files).")
+		listUsage()
+		os.Exit(1)
+	}
+	switch flag.Arg(1) {
+	case "users":
+		checkToken(token)
+		err := list.ListUsers(api_uri, token)
+		if err != nil {
+			fmt.Printf("Error: failed to get users, reason: %v\n", err)
+		}
```

Comment:
In the current setup, different errors will be followed by different help messages so that the help message is more specific relate to the errors. If we return all errors to the main function, the whole logic need to be rewritten. 
I can return just this error message without disturbing the logic but feels meaningless if we just return this error message but not others to the main function. 
I'm inclined to keep as is.

Reply:
You need to return an error to the main thread so that `os.Exit(1)` can be called from there and not the subfunction.

Reply:
I want specific error message and usage help for each command, e.g. for the following logic 
```golang
	if filepath == "" || username == "" {
		fmt.Println("Error: both -filepath and -user are required.")
		fileIngestUsage()
		os.Exit(1)
	}
```
If I return the error message, in the main function, it is not easy to track it is `fileIngestUsage` that should be printed and in the order after the error message. It's doable but will be a bit more complicated. 
If you think the problem is that os.Exit() should only happen in the main function. I can instead return the exit status code and then make sure os.Exit() happens only in the main function.

Reply:
Do you really need both the error message and the usage message?

Anyway this is a fully viable solution:
```
if filepath == "" || username == "" {
	fmt.Println("Error: both -filepath and -user are required.")
	fileIngestUsage()

	return fmt.Errorln("exit")
}
```